iceoryx2-pal/concurrency-sync/tests/iox_atomic_tests.rs

```diff
@@ -0,0 +1,619 @@
+    #[test]
+    fn fetch_sub_works<T: Req>() {
+        let n1 = T::generate_value();
+        let n2 = T::generate_value();
+
+        let sut = IoxAtomic::<T>::new(n1);
+
+        let result = sut.fetch_sub(n2, Ordering::Relaxed);
+
+        assert_that!(result, eq n1);
+        assert_that!(sut.load(Ordering::Relaxed), eq n1.overflowing_sub(n2).0);
+    }
+
+    fn ok_fetch_update<T: AddAssign + Copy>(value: T) -> Option<T> {
+        let mut temp = value;
+        temp += value;
+        Some(temp)
+    }
+
+    fn err_fetch_update<T: AddAssign + Copy>(_value: T) -> Option<T> {
+        None
+    }
+
+    #[test]
+    fn fetch_update_success_works<T: Req>() {
+        let n1 = T::generate_value();
+
+        let sut = IoxAtomic::<T>::new(n1);
+
+        let result = sut.fetch_update(Ordering::Relaxed, Ordering::Relaxed, ok_fetch_update::<T>);
+
+        assert_that!(result, is_ok);
+        assert_that!(result.unwrap(), eq n1);
+        let mut n = n1;
+        n += n1;
+        assert_that!(sut.load(Ordering::Relaxed), eq n);
+    }
+
+    #[test]
+    fn fetch_update_failure_works<T: Req>() {
+        let n1 = T::generate_value();
+
+        let sut = IoxAtomic::<T>::new(n1);
+
+        let result = sut.fetch_update(Ordering::Relaxed, Ordering::Relaxed, err_fetch_update::<T>);
+
+        assert_that!(result, is_err);
+        assert_that!(result.err().unwrap(), eq n1);
+        assert_that!(sut.load(Ordering::Relaxed), eq n1);
+    }
+
+    #[test]
+    fn fetch_xor_works<T: Req>() {
+        let n1 = T::generate_value();
+        let n2 = T::generate_value();
+
+        let sut = IoxAtomic::<T>::new(n1);
+
+        let result = sut.fetch_xor(n2, Ordering::Relaxed);
+
+        assert_that!(result, eq n1);
+        let mut bit_xor = n1;
+        bit_xor ^= n2;
+        assert_that!(sut.load(Ordering::Relaxed), eq bit_xor);
+    }
+
+    #[test]
+    fn into_inner_works<T: Req>() {
+        let n = T::generate_value();
+        let sut = IoxAtomic::<T>::new(n);
+
+        assert_that!(IoxAtomic::<T>::into_inner(sut), eq n);
+    }
+
+    #[test]
+    fn load_store_works<T: Req>() {
+        let n1 = T::generate_value();
+        let n2 = T::generate_value();
+        let sut = IoxAtomic::<T>::new(n1);
+
+        sut.store(n2, Ordering::Relaxed);
+
+        assert_that!(sut.load(Ordering::Relaxed), eq n2);
+    }
+
+    #[test]
+    fn swap_works<T: Req>() {
+        let n1 = T::generate_value();
+        let n2 = T::generate_value();
+        let sut = IoxAtomic::<T>::new(n1);
+
+        let result = sut.swap(n2, Ordering::Relaxed);
+
+        assert_that!(result, eq n1);
+        assert_that!(sut.load(Ordering::Relaxed), eq n2);
+    }
+
+    #[test]
+    fn compatibility_new_works<T: Req>() {
+        let n = T::generate_compatibility_value();
+        let sut = IoxAtomic::<T>::new(n);
+        let compat = AtomicU32::new(n.to_u32());
+
+        assert_that!(compat.load(Ordering::Relaxed), eq sut.load(Ordering::Relaxed).to_u32());
+    }
```

Comment:
The idea is to check if the `AtomicU32` behaves exactly like the `IoxAtomic`. Since the `IoxAtomic` covers everything from `u8` to `u128` I took `AtomicU32` as a representative - it is the largest atomic supported on 32-bit platforms. Furthermore, every atomic from `u8` to `u128` behaves identically when we ignore the integer size.